File: internal/service/expense_svc.go
package service

import (
	"context"
	"errors"
	"fmt"

	"github.com/zajca/zfaktury/internal/domain"
	"github.com/zajca/zfaktury/internal/repository"
)

// ExpenseService provides business logic for expense management.
type ExpenseService struct {
	repo  repository.ExpenseRepo
	audit *AuditService
}

// NewExpenseService creates a new ExpenseService.
func NewExpenseService(repo repository.ExpenseRepo, audit *AuditService) *ExpenseService {
	return &ExpenseService{repo: repo, audit: audit}
}

// Create validates and persists a new expense.
func (s *ExpenseService) Create(ctx context.Context, expense *domain.Expense) error {
	if expense.Description == "" {
		return errors.New("expense description is required")
	}
	if expense.Amount == 0 && len(expense.Items) == 0 {
		return errors.New("expense amount is required")
	}
	if expense.IssueDate.IsZero() {
		return errors.New("expense issue date is required")
	}
	if expense.CurrencyCode == "" {
		expense.CurrencyCode = domain.CurrencyCZK
	}
	if expense.BusinessPercent == 0 {
		expense.BusinessPercent = 100
	}
	if expense.BusinessPercent < 0 || expense.BusinessPercent > 100 {
		return errors.New("business share must be between 0 and 100")
	}

	// When items are present, recalculate totals from them.
	if len(expense.Items) > 0 {
		expense.CalculateTotals()
	} else if expense.VATAmount == 0 && expense.VATRatePercent > 0 {
		// Calculate VAT amount from rate if not set (flat-amount path).
		expense.VATAmount = expense.Amount.Multiply(float64(expense.VATRatePercent) / (100.0 + float64(expense.VATRatePercent)))
	}

	if err := s.repo.Create(ctx, expense); err != nil {
		return fmt.Errorf("creating expense: %w", err)
	}
	if s.audit != nil {
		s.audit.Log(ctx, "expense", expense.ID, "create", nil, expense)
	}
	return nil
}

// Update validates and updates an existing expense.
func (s *ExpenseService) Update(ctx context.Context, expense *domain.Expense) error {
	if expense.ID == 0 {
		return errors.New("expense ID is required")
	}
	if expense.Description == "" {
		return errors.New("expense description is required")
	}
	if expense.Amount == 0 && len(expense.Items) == 0 {
		return errors.New("expense amount is required")
	}
	if expense.BusinessPercent < 0 || expense.BusinessPercent > 100 {
		return errors.New("business share must be between 0 and 100")
	}

	// When items are present, recalculate totals from them.
	if len(expense.Items) > 0 {
		expense.CalculateTotals()
	} else if expense.VATAmount == 0 && expense.VATRatePercent > 0 {
		// Recalculate VAT amount from rate if not set (flat-amount path).
		expense.VATAmount = expense.Amount.Multiply(float64(expense.VATRatePercent) / (100.0 + float64(expense.VATRatePercent)))
	}

	existing, err := s.repo.GetByID(ctx, expense.ID)
	if err != nil {
		return fmt.Errorf("fetching expense for audit: %w", err)
	}
	if err := s.repo.Update(ctx, expense); err != nil {
		return fmt.Errorf("updating expense: %w", err)
	}
	if s.audit != nil {
		s.audit.Log(ctx, "expense", expense.ID, "update", existing, expense)
	}
	return nil
}

// Delete removes an expense by ID (soft delete).
func (s *ExpenseService) Delete(ctx context.Context, id int64) error {
	if id == 0 {
		return errors.New("expense ID is required")
	}
	if err := s.repo.Delete(ctx, id); err != nil {
		return fmt.Errorf("deleting expense: %w", err)
	}
	if s.audit != nil {
		s.audit.Log(ctx, "expense", id, "delete", nil, nil)
	}
	return nil
}

// GetByID retrieves an expense by its ID.
func (s *ExpenseService) GetByID(ctx context.Context, id int64) (*domain.Expense, error) {
	if id == 0 {
		return nil, errors.New("expense ID is required")
	}
	exp, err := s.repo.GetByID(ctx, id)
	if err != nil {
		return nil, fmt.Errorf("fetching expense: %w", err)
	}
	return exp, nil
}

// List retrieves expenses matching the given filter.
// Returns the expenses, total count, and any error.
func (s *ExpenseService) List(ctx context.Context, filter domain.ExpenseFilter) ([]domain.Expense, int, error) {
	if filter.Limit <= 0 {
		filter.Limit = 20
	}
	if filter.Limit > 100 {
		filter.Limit = 100
	}
	if filter.Offset < 0 {
		filter.Offset = 0
	}
	expenses, count, err := s.repo.List(ctx, filter)
	if err != nil {
		return nil, 0, fmt.Errorf("listing expenses: %w", err)
	}
	return expenses, count, nil
}

const maxBulkIDs = 500

// MarkTaxReviewed marks the given expense IDs as tax-reviewed.
func (s *ExpenseService) MarkTaxReviewed(ctx context.Context, ids []int64) error {
	if len(ids) == 0 {
		return errors.New("no expense IDs provided")
	}
	if len(ids) > maxBulkIDs {
		return errors.New("too many IDs, maximum is 500")
	}
	if err := s.repo.MarkTaxReviewed(ctx, dedupIDs(ids)); err != nil {
		return fmt.Errorf("marking expenses as tax reviewed: %w", err)
	}
	return nil
}

// UnmarkTaxReviewed removes the tax review mark from the given expense IDs.
func (s *ExpenseService) UnmarkTaxReviewed(ctx context.Context, ids []int64) error {
	if len(ids) == 0 {
		return errors.New("no expense IDs provided")
	}
	if len(ids) > maxBulkIDs {
		return errors.New("too many IDs, maximum is 500")
	}
	if err := s.repo.UnmarkTaxReviewed(ctx, dedupIDs(ids)); err != nil {
		return fmt.Errorf("unmarking expenses tax review: %w", err)
	}
	return nil
}

// dedupIDs removes duplicate int64 values preserving order.
func dedupIDs(ids []int64) []int64 {
	seen := make(map[int64]struct{}, len(ids))
	result := make([]int64, 0, len(ids))
	for _, id := range ids {
		if _, ok := seen[id]; !ok {
			seen[id] = struct{}{}
			result = append(result, id)
		}
	}
	return result
}

File: rust/crates/zfaktury-core/src/service/expense_svc.rs
use std::collections::HashSet;
use std::sync::Arc;

use zfaktury_domain::{Amount, CURRENCY_CZK, DomainError, Expense, ExpenseFilter};

use super::audit_svc::AuditService;
use crate::repository::traits::ExpenseRepo;

const MAX_BULK_IDS: usize = 500;

/// Service for expense management.
pub struct ExpenseService {
    repo: Arc<dyn ExpenseRepo + Send + Sync>,
    audit: Option<Arc<AuditService>>,
}

impl ExpenseService {
    pub fn new(repo: Arc<dyn ExpenseRepo + Send + Sync>, audit: Option<Arc<AuditService>>) -> Self {
        Self { repo, audit }
    }

    pub fn create(&self, expense: &mut Expense) -> Result<(), DomainError> {
        if expense.description.is_empty() {
            return Err(DomainError::InvalidInput);
        }
        if expense.amount == Amount::ZERO && expense.items.is_empty() {
            return Err(DomainError::InvalidInput);
        }
        if expense.issue_date == chrono::NaiveDate::default() {
            return Err(DomainError::InvalidInput);
        }
        if expense.currency_code.is_empty() {
            expense.currency_code = CURRENCY_CZK.to_string();
        }
        if expense.business_percent == 0 {
            expense.business_percent = 100;
        }
        if expense.business_percent < 0 || expense.business_percent > 100 {
            return Err(DomainError::InvalidInput);
        }

        if !expense.items.is_empty() {
            expense.calculate_totals();
        } else if expense.vat_amount == Amount::ZERO && expense.vat_rate_percent > 0 {
            expense.vat_amount = expense.amount.multiply(
                expense.vat_rate_percent as f64 / (100.0 + expense.vat_rate_percent as f64),
            );
        }

        self.repo.create(expense)?;
        if let Some(ref audit) = self.audit {
            audit.log("expense", expense.id, "create", None, None);
        }
        Ok(())
    }

    pub fn update(&self, expense: &mut Expense) -> Result<(), DomainError> {
        if expense.id == 0 || expense.description.is_empty() {
            return Err(DomainError::InvalidInput);
        }
        if expense.amount == Amount::ZERO && expense.items.is_empty() {
            return Err(DomainError::InvalidInput);
        }
        if expense.business_percent < 0 || expense.business_percent > 100 {
            return Err(DomainError::InvalidInput);
        }

        if !expense.items.is_empty() {
            expense.calculate_totals();
        } else if expense.vat_amount == Amount::ZERO && expense.vat_rate_percent > 0 {
            expense.vat_amount = expense.amount.multiply(
                expense.vat_rate_percent as f64 / (100.0 + expense.vat_rate_percent as f64),
            );
        }

        self.repo.get_by_id(expense.id)?; // verify exists
        self.repo.update(expense)?;
        if let Some(ref audit) = self.audit {
            audit.log("expense", expense.id, "update", None, None);
        }
        Ok(())
    }

    pub fn delete(&self, id: i64) -> Result<(), DomainError> {
        if id == 0 {
            return Err(DomainError::InvalidInput);
        }
        self.repo.delete(id)?;
        if let Some(ref audit) = self.audit {
            audit.log("expense", id, "delete", None, None);
        }
        Ok(())
    }

    pub fn get_by_id(&self, id: i64) -> Result<Expense, DomainError> {
        if id == 0 {
            return Err(DomainError::InvalidInput);
        }
        self.repo.get_by_id(id)
    }

    pub fn list(&self, mut filter: ExpenseFilter) -> Result<(Vec<Expense>, i64), DomainError> {
        if filter.limit <= 0 {
            filter.limit = 20;
        }
        if filter.limit > 100 {
            filter.limit = 100;
        }
        if filter.offset < 0 {
            filter.offset = 0;
        }
        self.repo.list(&filter)
    }

    pub fn mark_tax_reviewed(&self, ids: &[i64]) -> Result<(), DomainError> {
        if ids.is_empty() || ids.len() > MAX_BULK_IDS {
            return Err(DomainError::InvalidInput);
        }
        let deduped = dedup_ids(ids);
        self.repo.mark_tax_reviewed(&deduped)
    }

    pub fn unmark_tax_reviewed(&self, ids: &[i64]) -> Result<(), DomainError> {
        if ids.is_empty() || ids.len() > MAX_BULK_IDS {
            return Err(DomainError::InvalidInput);
        }
        let deduped = dedup_ids(ids);
        self.repo.unmark_tax_reviewed(&deduped)
    }
}

fn dedup_ids(ids: &[i64]) -> Vec<i64> {
    let mut seen = HashSet::new();
    ids.iter().copied().filter(|id| seen.insert(*id)).collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_dedup_ids() {
        assert_eq!(dedup_ids(&[1, 2, 2, 3, 1]), vec![1, 2, 3]);
        assert_eq!(dedup_ids(&[]), Vec::<i64>::new());
    }
}
